In [1]:
import pandas as pd

In [2]:
df = pd.read_csv("credit_card_fraud_2026.csv")

In [3]:
pd.set_option("display.max_column",None)

In [4]:
df.head()

,transaction_id,amount_usd,merchant_category,card_type,auth_method,channel,device_type,is_foreign_transaction,hours_since_last_txn,txn_count_last_24h,distance_from_home_km,card_age_months,customer_age,account_balance_usd,is_new_merchant,used_vpn,ip_country_mismatch,billing_shipping_mismatch,cvv_retry_count,velocity_score,time_of_day_hour,day_of_week,is_ai_generated_scam_attempt,merchant_risk_score,prior_disputes,is_fraud
0,1,42.86,Restaurants,Visa,OTP,Online,Android Phone,False,13.54,2,22.35,39,25,1080.71,False,False,False,False,0,0.1,18,3,False,42.3,0,0
1,2,4.75,Online Retail,Mastercard,3D Secure,Online,Android Phone,False,0.71,2,35.28,47,33,4463.72,True,False,False,False,0,25.8,12,4,False,28.3,0,0
2,3,77.18,Groceries,Mastercard,3D Secure,Online,Mac,False,0.35,5,9.82,51,59,366.25,False,True,False,True,0,42.3,5,0,False,24.7,1,0
3,4,1.69,Streaming,Visa,No Authentication,POS,Android Phone,False,3.42,6,19.72,48,28,1345.29,True,False,False,False,0,28.9,22,6,False,56.2,1,0
4,5,261.68,Travel,Visa,3D Secure,In-App,iPhone,False,2.43,2,17.68,43,36,4192.59,False,False,False,False,0,3.9,2,4,False,32.7,0,0


In [5]:
df.isna().sum()

transaction_id                  0
amount_usd                      0
merchant_category               0
card_type                       0
auth_method                     0
channel                         0
device_type                     0
is_foreign_transaction          0
hours_since_last_txn            0
txn_count_last_24h              0
distance_from_home_km           0
card_age_months                 0
customer_age                    0
account_balance_usd             0
is_new_merchant                 0
used_vpn                        0
ip_country_mismatch             0
billing_shipping_mismatch       0
cvv_retry_count                 0
velocity_score                  0
time_of_day_hour                0
day_of_week                     0
is_ai_generated_scam_attempt    0
merchant_risk_score             0
prior_disputes                  0
is_fraud                        0
dtype: int64

In [6]:
df.duplicated().sum()

np.int64(0)

In [7]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 20000 entries, 0 to 19999
Data columns (total 26 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   transaction_id                20000 non-null  int64  
 1   amount_usd                    20000 non-null  float64
 2   merchant_category             20000 non-null  object 
 3   card_type                     20000 non-null  object 
 4   auth_method                   20000 non-null  object 
 5   channel                       20000 non-null  object 
 6   device_type                   20000 non-null  object 
 7   is_foreign_transaction        20000 non-null  bool   
 8   hours_since_last_txn          20000 non-null  float64
 9   txn_count_last_24h            20000 non-null  int64  
 10  distance_from_home_km         20000 non-null  float64
 11  card_age_months               20000 non-null  int64  
 12  customer_age                  20000 non-null  int64  
 13  a

In [8]:
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import accuracy_score,precision_score,recall_score,f1_score
from sklearn.metrics import confusion_matrix
from sklearn.metrics import classification_report

In [9]:
x = df.drop(["transaction_id","is_fraud"],axis=1)
y = df["is_fraud"]

In [10]:
categorical_cols = [
    "merchant_category",
    "card_type",
    "auth_method",
    "channel",
    "device_type",
    "is_foreign_transaction",
    "is_new_merchant",
    "used_vpn",
    "ip_country_mismatch",
    "billing_shipping_mismatch",
    "is_ai_generated_scam_attempt"
]
x = pd.get_dummies(
    x,
    columns=categorical_cols,
    drop_first=True,
    dtype=int
)

In [11]:
x

,amount_usd,hours_since_last_txn,txn_count_last_24h,distance_from_home_km,card_age_months,customer_age,account_balance_usd,cvv_retry_count,velocity_score,time_of_day_hour,day_of_week,merchant_risk_score,prior_disputes,merchant_category_Electronics,merchant_category_Fuel,merchant_category_Gaming,merchant_category_Gift Cards,merchant_category_Groceries,merchant_category_Healthcare,merchant_category_Online Retail,merchant_category_Restaurants,merchant_category_Streaming,merchant_category_Travel,merchant_category_Utilities,card_type_Discover,card_type_Mastercard,card_type_RuPay,card_type_Visa,auth_method_Biometric,auth_method_No Authentication,auth_method_OTP,auth_method_PIN,channel_Contactless,channel_In-App,channel_Online,channel_POS,device_type_Android Phone,device_type_Mac,device_type_POS Terminal,device_type_Smart Watch,device_type_Tablet,device_type_Windows PC,device_type_iPhone,is_foreign_transaction_True,is_new_merchant_True,used_vpn_True,ip_country_mismatch_True,billing_shipping_mismatch_True,is_ai_generated_scam_attempt_True
0,42.86,13.54,2,22.35,39,25,1080.71,0,0.1,18,3,42.3,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,1,0,0,0,1,0,1,0,0,0,0,0,0,0,0,0,0,0,0
1,4.75,0.71,2,35.28,47,33,4463.72,0,25.8,12,4,28.3,0,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,1,0,0,0,0,0,0,0,1,0,0,0,0
2,77.18,0.35,5,9.82,51,59,366.25,0,42.3,5,0,24.7,1,0,0,0,0,1,0,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,1,0,0,0,0,0,0,0,1,0,1,0
3,1.69,3.42,6,19.72,48,28,1345.29,0,28.9,22,6,56.2,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,1,0,1,0,0,0,0,0,1,1,0,0,0,0,0,0,0,1,0,0,0,0
4,261.68,2.43,2,17.68,43,36,4192.59,0,3.9,2,4,32.7,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
19995,41.40,8.82,2,18.13,49,20,750.90,0,17.3,2,3,29.6,1,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,1,1,0,0,0,0,0,0,0,0,0,0,0,0
19996,47.92,45.81,1,8.97,41,74,2666.73,0,9.0,4,0,29.2,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,1,0,0,1,0,0,0,0,1,1,0,0,0,0,0,0,0,0,0,0,0,0
19997,10.25,20.67,0,48.67,40,65,284.25,1,1.0,19,5,22.9,0,0,0,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,1,0,0,0,0,0,0,1,0,0,0,0,0,0,0,0
19998,31.34,22.94,2,19.05,37,45,14609.39,0,8.8,2,1,27.9,0,0,0,0,0,0,0,0,0,0,1,0,0,0,0,1,0,0,0,0,0,0,1,0,1,0,0,0,0,0,0,0,0,0,0,0,0


In [12]:
x_train, x_test, y_train, y_test = train_test_split(
    x,y,test_size = 0.20, stratify=y, random_state=42
)

In [13]:
model = LogisticRegression( 
    max_iter=1000,
    class_weight="balanced")

In [14]:
model

LogisticRegression(class_weight='balanced', max_iter=1000)

In [15]:
model.fit(x_train, y_train)

C:\ProgramData\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:451: OptimizeWarning: Unknown solver options: iprint
  opt_res = optimize.minimize(
C:\ProgramData\anaconda3\Lib\site-packages\sklearn\linear_model\_logistic.py:465: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. OF ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  n_iter_i = _check_optimize_result(


LogisticRegression(class_weight='balanced', max_iter=1000)

In [16]:
y_pred = model.predict(x_test)

In [17]:
y_pred

array([1, 0, 0, ..., 0, 0, 0], shape=(4000,))

In [18]:
comparions = pd.DataFrame({
    "Actual": y_test,
    "Prediction": y_pred
})

In [19]:
comparions

,Actual,Prediction
12703,0,1
2670,0,0
15620,0,0
2931,0,0
5007,0,0
...,...,...
9964,0,0
13290,0,1
3197,0,0
18989,0,0


In [20]:
accuracy = accuracy_score(y_test,y_pred)*100
precision = precision_score(y_test,y_pred)
recall = recall_score(y_test,y_pred)*100
f1_score = f1_score(y_test,y_pred)*100

In [21]:
print(f"accuracy: {accuracy:.2f}%")
print("precision:",precision)
print("recall:",recall)
print("f1_score:",f1_score)

accuracy: 85.42%
precision: 0.09448818897637795
recall: 88.23529411764706
f1_score: 17.069701280227594


In [22]:
cm = confusion_matrix(y_test,y_pred)

In [23]:
cm

array([[3357,  575],
       [   8,   60]])

In [24]:
print(classification_report(y_test,y_pred))

              precision    recall  f1-score   support

           0       1.00      0.85      0.92      3932
           1       0.09      0.88      0.17        68

    accuracy                           0.85      4000
   macro avg       0.55      0.87      0.55      4000
weighted avg       0.98      0.85      0.91      4000



The Logistic Regression model with balanced class weights achieved 85.42% accuracy and 88.24% recall for fraud detection. It successfully identified 60 out of 68 fraudulent transactions, although the low precision of 9.45% indicates a relatively high false-positive rate.